bfd/cpu-arc.c

```diff
@@ -53,3 +56,54 @@ static const bfd_arch_info_type arch_info_struct[] =
 
 const bfd_arch_info_type bfd_arc_arch =
   ARC (bfd_mach_arc_arc600, "ARC600", TRUE, &arch_info_struct[0]);
+
+/* ARC-specific "compatible" function.  The general rule is that if A and B are
+   compatible, then this function should return architecture that is more
+   "feature-rich", that is, can run both A and B.  ARCv2, EM and HS all has
+   same mach number, so bfd_default_compatible assumes they are the same, and
+   returns an A.  That causes issues with GDB, because GDB assumes that if
+   machines are compatible, then "compatible ()" always returns same machine
+   regardless of argument order.  As a result GDB gets confused because, for
+   example, compatible (ARCv2, EM) returns ARCv2, but compatible (EM, ARCv2)
+   returns EM, hence GDB is not sure if they are compatible and prints a
+   warning.  */
+
+static const bfd_arch_info_type *
+arc_compatible (const bfd_arch_info_type *a, const bfd_arch_info_type *b)
+{
+  const bfd_arch_info_type * const arc600 = &bfd_arc_arch;
+  const bfd_arch_info_type * const arc700 = &arch_info_struct[2];
+  const bfd_arch_info_type * const em = &arch_info_struct[5];
+  const bfd_arch_info_type * const hs = &arch_info_struct[6];
+
+  /* If a & b are for different architecture we can do nothing.  */
+  if (a->arch != b->arch)
+    return NULL;
+
+  if (a->bits_per_word != b->bits_per_word)
+    return NULL;
+
+  /* HS and EM are not compatible.  */
+  if ((a == em && b == hs) || (a == hs && b == em))
+    return NULL;
+
+  /* ARCv2|EM and EM.  */
+  if ((a->mach == bfd_mach_arc_arcv2 && b == em)
+      || (b->mach == bfd_mach_arc_arcv2 && a == em))
+    return em;
+
+  /* ARCv2|HS and HS.  */
+  if ((a->mach == bfd_mach_arc_arcv2 && b == hs)
+      || (b->mach == bfd_mach_arc_arcv2 && a == hs))
+    return hs;
+
+  /* ARC700.  */
+  if (a->mach == bfd_mach_arc_arc700 && b->mach == bfd_mach_arc_arc700)
+    return arc700;
+
+  /* ARC600.  */
+  if (a->mach == bfd_mach_arc_arc600 && b->mach == bfd_mach_arc_arc600)
+    return arc600;
+
```

Comment:
Hi,
we need to handle also the bfd_mach_arc_arc601, and come up with something for the old a4 and a5 configurations (i.e., bfd_mach_arc_a4, and bfd_mach_arc_a5). Otherwise it is alright.